### PR TITLE
Must support encryption at rest

### LINKED ISSUE #113: Must support encryption at rest
## Feature description
Encrypt all objects written to persistent storage using Vault as KMS

## Why is it needed?
Multiple requirements related to security, chief amonst which are Separation of Duties, GDPR and national directives for storing personal information.

## Checklist (after created issue)
- [ ] Added label(s)
- [ ] Added to project
- [ ] Added to milestone

### PR DESCRIPTION
**Pull request by dangernil**
## Closes/fixes/resolves issue(s)?
Closes #113

## What was added/changed/fixed?
- Added Native Vault KMS support using Vault transit encryption #113
  - In this version we are using the deprecated native vault integration.
  - EOL for this feature is October 2021
  - Switch to [MinIO KES](https://github.com/minio/kes/wiki/Hashicorp-Vault-Keystore) before then

## Checklist (after created PR)
- [ ] Added reviewers
- [ ] Added assignee(s)
- [ ] Added label(s)
- [ ] Added to project
- [ ] Added to milestone

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -144,7 +146,10 @@ module "minio" {
 | vault_secret.vault_kv_field_secret_key | Secret key name in Vault kv path | string | "secret_key" |
 | minio\_upstreams | List up connect upstreams | list(object) | [] | no |
 | mc\_extra\_commands | Extra commands to run in MC container after creating buckets | list(string) | [] | no |
-
```
Comment (claesgill) [inline README.md:148]: Think we also need a section about the KMS and how to use it.

File: conf/nomad/minio.hcl [hunk 2]
```diff
@@ -137,6 +137,24 @@ MINIO_SECRET_KEY="${secret_key}"
 ${ envs }
 EOF
       }
+%{ if use_vault_kms }
+template {
```
Comment (claesgill) [inline conf/nomad/minio.hcl]: Think this should have the same indentation as the other templates

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+version: 3
```
Comment (claesgill) [inline docker-compose.yml:1]: Dont think you need this file anymore?

Comment (dangernil) [reply docker-compose.yml:1]: Tested it, and we do.

File: conf/nomad/minio.hcl (hunk 2 above)
Comment (dangernil) [reply conf/nomad/minio.hcl]: Something like the updated one?

File: conf/nomad/minio.hcl [hunk 4]
```diff
@@ -125,18 +125,37 @@ job "${service_name}" {
...
+      data = <<EOH
+        {{ with secret "${vault_kms_approle_kv}" }}
+          MINIO_KMS_VAULT_APPROLE_ID="{{ .Data.data.approle_id }}"
+          MINIO_KMS_VAULT_APPROLE_SECRET="{{ .Data.data.secret_id }}"
+          {{end}}
+          MINIO_KMS_VAULT_ENDPOINT=${vault_address}
+          MINIO_KMS_VAULT_KEY_NAME=${vault_kms_key_name}
+          MINIO_KMS_VAULT_AUTH_TYPE=approle
+          MINIO_KMS_AUTO_ENCRYPTION=on
+          MINIO_KMS_VAULT_DEPRECATION=off
```
Comment (claesgill) [inline conf/nomad/minio.hcl:157]: Should these always be static - never changed?

File: conf/nomad/minio.hcl [hunk 5]
```diff
@@ -125,18 +125,37 @@ job "${service_name}" {
...
+        EOH
+
+        destination = "secrets/kms"
```
Comment (claesgill) [inline conf/nomad/minio.hcl:160]: Same question as above

File: variables.tf [hunk 6]
```diff
@@ -167,3 +167,27 @@ variable "mc_extra_commands" {
...
+  type = string
+  default = ""
+  description = "Path to key in vault where ApproleID and SecretID is stored. Only relevant when Vault KMS is used."
+}
+
+variable "vault_kms_key_name" {
+  type = string
+  default = ""
+  description = "Name of key in vault transit engine. Only relevant when Vault KMS is used."
+}
```
Comment (claesgill) [inline variables.tf]: I think it's better to wrap these in a nested variable. Similar to the `vault_secret` variable. WDYT?

File: docker-compose.yml (hunk 3 above)
Comment (claesgill) [reply docker-compose.yml:1]: Ok. Can you point me to the code that triggers docker-compose? Cant find where this file is run from.

Also, this file should be in the `docker`-folder.

File: variables.tf (hunk 6 above)
Comment (dangernil) [reply variables.tf]: Yeah I can do that, it´s not that many variables, but if you think it´ll look cleaner with them wrapped that´s fine.

File: conf/nomad/minio.hcl (hunk 4 above)
Comment (dangernil) [reply conf/nomad/minio.hcl:157]: For now I think its better to have them as static. But can look into other options to type, encryption etc later, but not sure we ever want to change it to be honest.

File: conf/nomad/minio.hcl (hunk 5 above)
Comment (dangernil) [reply conf/nomad/minio.hcl:160]: It´s set as static for simplicity I believe.

File: variables.tf (hunk 6 above)
Comment (claesgill) [reply variables.tf]: I think it makes sense to collect them in one variable because since all should be used in the same context.

Comment (dangernil) [reply variables.tf]: It did not like variables to be nested when there is a empty default string, so leaving it as is.

Comment (claesgill) [reply variables.tf]: I don't see the problem with a nested variable. Can you ellaborate?

**Review by zhenik — DISMISSED** [review]
LGTM

File: README.md (hunk 1 above)
Comment (claesgill) [reply README.md:148]: Still need doc

**Review by claesgill — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -216,7 +221,16 @@ module "minio" {
                   }
 }
 ```
-
+### Key Management Secrets (KMS)
+The Key Management secrets engine provides a consistent workflow for distribution and lifecycle management of cryptographic keys in various key management service (KMS) providers.
```
Comment (claesgill) [inline README.md:225]: There are 2 other variables that should be mentioned in this context (`vault_address`, `vault_kms_approle_kv`).

Also, think that the example should show implementation not one of the variables.

Comment (claesgill) [inline README.md]: There are 2 other variables that should be mentioned in this context (`vault_address`, `vault_kms_approle_kv`).

Also, think that the example should show implementation not one of the variables.

**Review by claesgill — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -216,6 +221,35 @@ module "minio" {
...
+    }
+  EOT
+  path = "secret/kms"
+}
+```
+
+``use_vault_kms``
+This is false by default, but can be turned on if you want to use vaults integrated transit encryption to manage your keys.
+The keys will then be store in ``secrets/kms`` folder inside of vault. You can change the path where the keys
+are stored by changing this variable ``vault_kms_approle_kv`` but that is only relevant if you `use_vault_kms = true`.
```
Comment (claesgill) [inline README.md:270]: It's not clear to me how to use this. Could you also add a minimal example of usage?

Something like:

```hcl
resource "vault_generic_secret" "kms_transit_key" {
   data_json = "{}"
   path = "transit/keys/minio"
}
resource "vault_generic_secret" "kms_approle" {
  data_json = <<EOT
    {
      "approle_id": "${vault_approle_auth_backend_role.minio_kms.role_id}" ,
      "secret_id": "${vault_approle_auth_backend_role_secret_id.minio_kms.secret_id}"
    }
  EOT
  path = "secret/kms"
}

module minio {
    # ... other configuration

    use_vault_kms = true
    # Other variables need to be set here?
}

File: variables.tf (hunk 6 above)
Comment (claesgill) [reply variables.tf]: bump

**Review by claesgill — DISMISSED** [review]

**Review by claesgill — APPROVED** [review]

**Review by zhenik — APPROVED** [review]
LGTM

**Pull request merged**